### ISSUE TITLE
On Linux: An error occurred trying to start process '/usr/local/share/dotnet/dotnet' with working directory '/builds'. No such file or directory

### ISSUE THREAD

**Issue by j-oxborrow** [CycloneDX/cyclonedx-dotnet#806]
Hello,

With the update to CycloneDX 3.0.0 this morning the creation of SBOMs in our CI/CD began to fail.

We are using the Dotnet 6 SDK Docker image from Microsoft from available here [mcr.microsoft.com/dotnet/sdk:6.0](url)

We are starting the analysis using the command:
`dotnet CycloneDX *.csproj -sv <project_name>.csproj --recursive -tfm net6.0 --runtime linux-x64 -f sbom.xml -o ./sbom/`

The analysis immediately fails with an unhandled exception, the stack trace is:

```Unhandled exception: System.ComponentModel.Win32Exception (2): An error occurred trying to start process '/usr/local/share/dotnet/dotnet' with working directory '/build'. No such file or directory
   at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
   at CycloneDX.Services.DotnetCommandService.Run(String workingDirectory, String arguments) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetCommandService.cs:line 52
   at CycloneDX.Services.DotnetCommandService.Run(String arguments) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetCommandService.cs:line 36
   at CycloneDX.Services.DotnetUtilsService.GetGlobalPackagesCacheLocation() in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetUtilsService.cs:line 72
   at CycloneDX.Services.DotnetUtilsService.GetPackageCachePaths() in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetUtilsService.cs:line 108
   at CycloneDX.Runner.HandleCommandAsync(RunOptions options) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Runner.cs:line 95
   at CycloneDX.Program.<>c__DisplayClass0_0.<<Main>b__2>d.MoveNext() in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Program.cs:line 137
--- End of stack trace from previous location ---
   at System.CommandLine.Invocation.AnonymousCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.AnonymousCommandHandler.SyncUsingAsync(InvocationContext context)
   at System.CommandLine.Invocation.AnonymousCommandHandler.Invoke(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass17_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass12_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseVersionOption>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass19_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__18_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__5_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass8_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

The directory does exist as that is where I am running the command from.

I've attempted to update the arguments used in the command above to the new arguments as outlined in the v3.0.0 release notes [https://github.com/CycloneDX/cyclonedx-dotnet/releases](url) here, but receive the same stack trace and error.

Uninstalling v3.0.0 and installing v2.10.0 creates the SBOM as expected.

I've recreated the issue in multiple Dotnet 6.0 projects.

Please let me know what else I can provide to help with this issue.

**Comment by mtsfoni** [comment]
Hello,

thank you for reporting, and sorry for the inconvenience.

It's a little mysterious, the error comes from a class that I didn't touch. However I will of course try to find a solution quickly.

Apparently it crashes when cdx tries to call `nuget locals global-packages --list`

What exception is thrown? Is it a directory not found exception?

**Comment by j-oxborrow** [comment]
Thanks for the quick reply. I didn't notice that I cut off the top of the exception trace. It is a directory not found error. The full trace is below.
```
Unhandled exception: System.ComponentModel.Win32Exception (2): An error occurred trying to start process '/usr/local/share/dotnet/dotnet' with working directory '/builds'. No such file or directory
   at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
   at CycloneDX.Services.DotnetCommandService.Run(String workingDirectory, String arguments) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetCommandService.cs:line 52
   at CycloneDX.Services.DotnetCommandService.Run(String arguments) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetCommandService.cs:line 36
   at CycloneDX.Services.DotnetUtilsService.GetGlobalPackagesCacheLocation() in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetUtilsService.cs:line 72
   at CycloneDX.Services.DotnetUtilsService.GetPackageCachePaths() in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetUtilsService.cs:line 108
   at CycloneDX.Runner.HandleCommandAsync(RunOptions options) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Runner.cs:line 95
   at CycloneDX.Program.<>c__DisplayClass0_0.<<Main>b__2>d.MoveNext() in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Program.cs:line 137
--- End of stack trace from previous location ---
   at System.CommandLine.Invocation.AnonymousCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.AnonymousCommandHandler.SyncUsingAsync(InvocationContext context)
   at System.CommandLine.Invocation.AnonymousCommandHandler.Invoke(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass17_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass12_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseVersionOption>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass19_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__18_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__5_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass8_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

**Comment by mtsfoni** [comment]
I can reproduce the problem and probably found the responsible change.
Working on a fix.

**Comment by mtsfoni** [comment]
Should be fixed in 3.0.1

I leave this open in case other people have the same problem and for further investigation and learnings.
This is one of those bugs where one looks at and says, "of course that doesn't work, but how did that ever work before?"

Anyway, it's 2 am; good night.

**Comment by j-oxborrow** [comment]
Thanks Michael,

I re-ran a failed CI/CD pipeline and it pulled the v3.0.1 which ran successfully.

Jordan Oxborrow

Automation Engineer
Sedna Inc | www.sedna.net<http://www.sedna.net/>
US Cell/Signal: +1 775 293 0735

________________________________

**Comment by a-korevaar** [comment]
Running into the same issue with v3.0.1 from an Azure DevOps agent running in a Docker container. I've checked the working directory and path to the solution to be correct.

```
2023-12-29T06:47:31.3998753Z [command]/usr/bin/dotnet tool install CycloneDX --tool-path /azp/_work/_temp/e57a2e9a-7499-4190-94fc-df4ac314655a
2023-12-29T06:47:36.2292709Z You can invoke the tool using the following command: dotnet-CycloneDX
2023-12-29T06:47:36.2293802Z Tool 'cyclonedx' (version '3.0.1') was successfully installed.
2023-12-29T06:47:36.2400845Z Running CycloneDX for /azp/_work/1/s/src/***.sln
2023-12-29T06:47:36.2412327Z [command]/azp/_work/_temp/e57a2e9a-7499-4190-94fc-df4ac314655a/dotnet-CycloneDX /azp/_work/1/s/src/***.sln --out /azp/_work/1/a/deployment --json --disable-github-licenses
2023-12-29T06:47:36.5293323Z Unhandled exception: System.ComponentModel.Win32Exception (2): An error occurred trying to start process '/usr/lib/dotnet/dotnet' with working directory '/azp/_work/1/s'. No such file or directory
2023-12-29T06:47:36.5296542Z    at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec) in /_/src/runtime/artifacts/source-build/self/src/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:line 546
2023-12-29T06:47:36.5298879Z    at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo) in /_/src/runtime/artifacts/source-build/self/src/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:line 456
2023-12-29T06:47:36.5300726Z    at System.Diagnostics.Process.Start(ProcessStartInfo startInfo) in /_/src/runtime/artifacts/source-build/self/src/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:line 1350
2023-12-29T06:47:36.5302555Z    at CycloneDX.Services.DotnetCommandService.Run(String workingDirectory, String arguments) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetCommandService.cs:line 52
2023-12-29T06:47:36.5304232Z    at CycloneDX.Services.DotnetCommandService.Run(String arguments) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetCommandService.cs:line 36
2023-12-29T06:47:36.5306009Z    at CycloneDX.Services.DotnetUtilsService.GetGlobalPackagesCacheLocation() in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetUtilsService.cs:line 72
2023-12-29T06:47:36.5307975Z    at CycloneDX.Services.DotnetUtilsService.GetPackageCachePaths() in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/DotnetUtilsService.cs:line 108
2023-12-29T06:47:36.5310013Z    at CycloneDX.Runner.HandleCommandAsync(RunOptions options) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Runner.cs:line 95
2023-12-29T06:47:36.5311464Z    at CycloneDX.Program.<>c__DisplayClass0_0.<<Main>b__2>d.MoveNext() in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Program.cs:line 137
2023-12-29T06:47:36.5312463Z --- End of stack trace from previous location ---
2023-12-29T06:47:36.5313180Z    at System.CommandLine.Invocation.AnonymousCommandHandler.InvokeAsync(InvocationContext context)
2023-12-29T06:47:36.5313992Z    at System.CommandLine.Invocation.AnonymousCommandHandler.SyncUsingAsync(InvocationContext context)
2023-12-29T06:47:36.5314959Z    at System.CommandLine.Invocation.AnonymousCommandHandler.Invoke(InvocationContext context)
2023-12-29T06:47:36.5315805Z    at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
2023-12-29T06:47:36.5316719Z --- End of stack trace from previous location ---
2023-12-29T06:47:36.5317488Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass17_0.<<UseParseErrorReporting>b__0>d.MoveNext()
2023-12-29T06:47:36.5318404Z --- End of stack trace from previous location ---
2023-12-29T06:47:36.5319119Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass12_0.<<UseHelp>b__0>d.MoveNext()
2023-12-29T06:47:36.5320015Z --- End of stack trace from previous location ---
2023-12-29T06:47:36.5320870Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseVersionOption>b__0>d.MoveNext()
2023-12-29T06:47:36.5321781Z --- End of stack trace from previous location ---
2023-12-29T06:47:36.5322533Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass19_0.<<UseTypoCorrections>b__0>d.MoveNext()
2023-12-29T06:47:36.5323455Z --- End of stack trace from previous location ---
2023-12-29T06:47:36.5324172Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__18_0>d.MoveNext()
2023-12-29T06:47:36.5325515Z --- End of stack trace from previous location ---
2023-12-29T06:47:36.5326529Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseDirective>b__0>d.MoveNext()
2023-12-29T06:47:36.5327530Z --- End of stack trace from previous location ---
2023-12-29T06:47:36.5328524Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__5_0>d.MoveNext()
2023-12-29T06:47:36.5329433Z --- End of stack trace from previous location ---
2023-12-29T06:47:36.5330194Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass8_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

**Comment by mtsfoni** [comment]
> I noticed the code [tries a few paths](https://github.com/CycloneDX/cyclonedx-dotnet/blob/34cc4e5aa6e9781e12513f567ee52e836e053648/CycloneDX/Services/DotnetCommandService.cs#L116) to find dotnet. Is there a cleaner way to resolve/derive the path, such as using the PATH variable? I guess we could also set the DOTNET_ROOT environment variable to point to the correct directory.

Setting DOTNET_ROOT might work as a workaround. 
The weird thing is, the function that tries to locate dotnet worked before as part of a package. 
Trying multiple paths was only meant as a band-aid, to quickly fix the release.

**Comment by a-korevaar** [comment]
Not sure, but [e39864c](https://github.com/CycloneDX/cyclonedx-dotnet/commit/e39864cd7780caba3c345356f29ef392f5d2950b) shows it has been a rather simple implementation since at least two months?

```
// DOTNET_ROOT specifies the location of the .NET runtimes, if they are not installed in the default location.
var dotnetRoot = Environment.GetEnvironmentVariable("DOTNET_ROOT");

if (string.IsNullOrEmpty(dotnetRoot))
{
    // fall back to default location
    // https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-environment-variables#dotnet_root-dotnet_rootx86
    dotnetRoot = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "C:\\Program Files\\dotnet" : "/usr/local/share/dotnet";
}
```

It's not using any lib to get the path, just a ternary with a fixed path for Windows or any other OS.

**Comment by mtsfoni** [comment]
I found the source of the problem. 
I first make a new release, then I'll explain how it came to this bug.

**Comment by mtsfoni** [comment]
There is a version 3.0.2 now, which should work the same way (in this aspect) as 2.10.0 did.

The story is quite an adventure:

A contributor replaced McMasterCommandLineUtils with System.CommandLine, and initially, everything seemed to be working well. However, there was a function in the McMaster package that we used to resolve the dotnet path. To address this, the contributor copied the code of that specific function, allowing us to eliminate the McMaster package. Initially, it appeared to be a successful move.

However, there was a catch. Up until then, we had been using Package-Version 4.0.2, while the latest version was 4.1.0. In their latest update, the maintainers had actually modified the very code we had integrated. The problem arose because the altered code no longer functioned as expected. Instead of giving up and returning 'dotnet' as in the older version, it consistently returned a folder. This was a critical issue as our previous implementation relied on the assumption that the code would simply return 'dotnet,' leaving the handling to the PATH environment variable.

Adding a humorous twist to the situation, this bug would have surfaced even if we hadn't switched to System.CommandLine but had instead updated McMasterCommandLineUtils to 4.10.0. In that case, it could have been introduced by DependaBot.

This incident now finds its place in my ledger of war stories, and one day, I'll regale my nieces and nephews with the tale.
